### PR TITLE
Fix infinite loop in AzureObjectStorage::listObjects

### PR DESCRIPTION
**Pull request by jkartseva**
While the exit condition is correct for `AzureIteratorAsync`, listObject may never exit.

This issue reproduces when passing an empty path.
Tested with the following integration test:
https://pastila.nl/?01691b54/9ceaf103c91ab9e38ea3541abfcd1ae6#hUtrrz5sFHcBSHCLS9AlVA== (will be added when plain_rewritable is enabled for all disk types).

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix infinite loop while listing objects in Azure blob storage

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 9607ba9f911f3a0398bacb81c6e4fa3ac49e0e3b with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63257/9607ba9f911f3a0398bacb81c6e4fa3ac49e0e3b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp [hunk 1]
```diff
@@ -193,8 +193,7 @@ void AzureObjectStorage::listObjects(const std::string & path, RelativePathsWith
                 break;
             options.PageSizeHint = keys_left;
         }
-
-        if (blob_list_response.HasPage())
+        if (blob_list_response.NextPageToken.HasValue() && !blob_list_response.NextPageToken.Value().empty())
```
Comment (alesapin) [inline src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:196]: I've found this way to check in example https://github.com/Azure/azure-sdk-for-cpp/blob/azure-storage-blobs_12.10.0/sdk/storage/azure-storage-blobs/samples/blob_list_operation.cpp#L53-L71. Is it correct one?

Comment (jkartseva) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:196]: I think it is correct under the condition that `MoveToNextPage()` is used as an iterator:
``` cpp
for (auto containerPage = serviceClient.ListBlobContainers(); containerPage.HasPage();
       containerPage.MoveToNextPage())
```

It doesn't work with `options.ContinuationToken = blob_list_response.NextPageToken;`

Comment (jkartseva) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:196]: Checking with HasPage(), MoveToNextPage()

Comment (jkartseva) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp]: So the issue is that `BlobContainerClient::ListBlobs` sets a page, thus the sequential `blob.HasPage()` condition is `true`:
```
2024.05.03 01:43:16.438300 [ 8 ] {} <Trace> AzureObjectStorage before ListBlobs: HasPage: false
2024.05.03 01:43:16.440368 [ 8 ] {} <Trace> AzureObjectStorage after ListBlobs: HasPage: true
```
And the loop never breaks...

I modified the fix so it follows the example. Namely, the condition `HasPage()` is verified before listing blobs.

**Review by alesapin — APPROVED** [review]

**Pull request merged**